File: src/main/java/net/guizhanss/fastmachines/libs/guizhanlib/slimefun/items/builder/RecipeBuilder.java
package net.guizhanss.fastmachines.libs.guizhanlib.slimefun.items.builder;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import org.bukkit.Material;
import org.bukkit.inventory.ItemStack;

import io.github.thebusybiscuit.slimefun5.api.items.SlimefunItemStack;
import io.github.thebusybiscuit.slimefun5.libraries.xseries.XMaterial;

import net.guizhanss.fastmachines.utils.MaterialCompat;

/**
 * A small builder for describing a 3x3 shaped recipe as three 3-character rows ({@link #row}) plus a
 * mapping from each character to an ingredient ({@link #where}). A blank space always means
 * "no ingredient".
 *
 * Java-8-safe port of GuizhanLib-kt's {@code buildRecipe} DSL.
 */
public class RecipeBuilder {

    private final List<String> rows = new ArrayList<>();
    private final Map<Character, ItemStack> ingredients = new HashMap<>();

    public static RecipeBuilder create() {
        return new RecipeBuilder();
    }

    public RecipeBuilder row(String row) {
        if (row.length() != 3) {
            throw new IllegalArgumentException("Each recipe row must be exactly 3 characters long, was \"" + row + "\"");
        }
        rows.add(row);
        return this;
    }

    /** Version-safe ingredient: the {@link XMaterial} is resolved to a {@link Material} that exists on this server. */
    public RecipeBuilder where(char c, XMaterial material) {
        ingredients.put(c, material == null ? null : new ItemStack(MaterialCompat.safe(material)));
        return this;
    }

    public RecipeBuilder where(char c, Material material) {
        ingredients.put(c, material == null ? null : new ItemStack(material));
        return this;
    }

    public RecipeBuilder where(char c, SlimefunItemStack item) {
        ingredients.put(c, item == null ? null : item.item());
        return this;
    }

    public RecipeBuilder where(char c, ItemStack item) {
        ingredients.put(c, item);
        return this;
    }

    public ItemStack[] build() {
        if (rows.size() != 3) {
            throw new IllegalArgumentException("A recipe must have exactly 3 rows, had " + rows.size());
        }
        String shape = rows.get(0) + rows.get(1) + rows.get(2);
        ItemStack[] recipe = new ItemStack[9];
        for (int i = 0; i < 9; i++) {
            recipe[i] = ingredients.get(shape.charAt(i));
        }
        return recipe;
    }
}

File: src/main/kotlin/net/guizhanss/fastmachines/libs/guizhanlib/slimefun/items/builder/RecipeBuilder.kt
package net.guizhanss.fastmachines.libs.guizhanlib.slimefun.items.builder

import io.github.thebusybiscuit.slimefun5.api.items.SlimefunItemStack
import io.github.thebusybiscuit.slimefun5.libraries.xseries.XMaterial
import net.guizhanss.fastmachines.utils.MaterialCompat
import org.bukkit.Material
import org.bukkit.inventory.ItemStack

/**
 * A small DSL for describing a 3x3 shaped recipe as three 3-character rows (`+"..."`) plus a mapping
 * from each character to an ingredient (`'X' means ...`). A blank space always means "no ingredient".
 *
 * Java-8-safe, Kotlin port of GuizhanLib-kt's `buildRecipe` DSL - see the package-level note in
 * [net.guizhanss.fastmachines.libs.guizhanlib].
 */
class RecipeBuilder {

    private val rows = mutableListOf<String>()
    private val ingredients = mutableMapOf<Char, ItemStack?>()

    operator fun String.unaryPlus() {
        require(length == 3) { "Each recipe row must be exactly 3 characters long, was \"$this\"" }
        rows += this
    }

    infix fun Char.means(material: Material?) {
        ingredients[this] = material?.let { ItemStack(it) }
    }

    /** Version-safe ingredient: the [XMaterial] is resolved to a [Material] that exists on this server. */
    infix fun Char.means(material: XMaterial?) {
        ingredients[this] = material?.let { ItemStack(MaterialCompat.safe(it)) }
    }

    infix fun Char.means(item: SlimefunItemStack?) {
        ingredients[this] = item?.item()
    }

    infix fun Char.means(item: ItemStack?) {
        ingredients[this] = item
    }

    internal fun build(): Array<ItemStack?> {
        require(rows.size == 3) { "A recipe must have exactly 3 rows, had ${rows.size}" }
        val shape = rows.joinToString("")
        return Array(9) { i -> ingredients[shape[i]] }
    }
}

fun buildRecipe(block: RecipeBuilder.() -> Unit): Array<ItemStack?> {
    val builder = RecipeBuilder()
    builder.block()
    return builder.build()
}
